File: myintent/src/main/java/pl/mareklangiewicz/myintent/MIContract.java
package pl.mareklangiewicz.myintent;

import android.app.SearchManager;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.BaseColumns;

import java.util.List;

import pl.mareklangiewicz.myutils.MyCommands;

import static pl.mareklangiewicz.myutils.MyTextUtils.str;

/**
 * Created by Marek Langiewicz on 07.10.15.
 */
public final class MIContract {

    public static final String AUTH = "pl.mareklangiewicz.myintent.provider";
    public static final Uri BASE_URI = Uri.parse("content://" + AUTH);

    private MIContract() {
        throw new AssertionError("MIContract class is noninstantiable.");
    }

    public static final class CmdRecent implements BaseColumns {

        public static final String PATH = "cmd/recent";

        public static final Uri URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build();

        public static final String TYPE_DIR = ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + AUTH + "/" + PATH;
        public static final String TYPE_ITEM = ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + AUTH + "/" + PATH;

        public static final String TABLE_NAME = CmdRecent.class.getSimpleName();

        public static final String COL_COMMAND = "Command";
        public static final String COL_TIME = "Time";

        public static int clear(Context context) {
            ContentResolver cr = context.getContentResolver();
            return cr.delete(URI, null, null);
        }

        public static Uri insert(Context context, String command) {
            ContentResolver cr = context.getContentResolver();
            ContentValues values = new ContentValues();
            values.put(COL_COMMAND, command);
            values.put(COL_TIME, System.currentTimeMillis());
            return cr.insert(URI, values);
        }

        /**
         * limit <= 0 means no limit
         * returns true if success
         */
        public static boolean load(Context context, List<String> commands, int limit) {

            String slimit = limit <= 0 ? "" : " LIMIT " + str(limit) + " ";
            ContentResolver cr = context.getContentResolver();
            Cursor c = cr.query(URI, null, null, null, " " + COL_TIME + " DESC " + slimit);
            if(c == null)
                return false;
            boolean empty = !c.moveToFirst();
            if(empty) {
                c.close();
                return true;
            }
            do commands.add(c.getString(c.getColumnIndex(COL_COMMAND)));
            while(c.moveToNext());
            c.close();
            return true;
        }

    }


    public static final class CmdExample implements BaseColumns {

        public static final String PATH = "cmd/example";

        public static final Uri URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build();

        public static final String TYPE_DIR = ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + AUTH + "/" + PATH;
        public static final String TYPE_ITEM = ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + AUTH + "/" + PATH;

        public static final String TABLE_NAME = CmdExample.class.getSimpleName();

        public static final String COL_COMMAND = "Command";
        public static final String COL_PRIORITY = "Priority";

        public static int clear(Context context) {
            ContentResolver cr = context.getContentResolver();
            return cr.delete(URI, null, null);
        }

        public static Uri insert(Context context, String command, long priority) {
            ContentResolver cr = context.getContentResolver();
            ContentValues values = new ContentValues();
            values.put(COL_COMMAND, command);
            values.put(COL_PRIORITY, priority);
            return cr.insert(URI, values);
        }

        public static boolean load(Context context, List<String> commands) {
            ContentResolver cr = context.getContentResolver();
            Cursor c = cr.query(URI, null, null, null, " " + COL_PRIORITY + " DESC ");
            if(c == null)
                return false;
            boolean empty = !c.moveToFirst();
            if(empty) {
                c.close();
                return true;
            }
            do commands.add(c.getString(c.getColumnIndex(COL_COMMAND)));
            while(c.moveToNext());
            c.close();
            return true;
        }
    }


    public static final class CmdSuggest implements BaseColumns {

        public static final String PATH = "cmd/" + SearchManager.SUGGEST_URI_PATH_QUERY;

        public static final Uri URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build();

        public static final String TYPE_DIR = ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + AUTH + "/" + PATH;

        public static final String TABLE_NAME = CmdSuggest.class.getSimpleName();

        public static final String COL_QUERY = SearchManager.SUGGEST_COLUMN_QUERY;
        public static final String COL_TEXT = SearchManager.SUGGEST_COLUMN_TEXT_1;
        public static final String COL_ICON = SearchManager.SUGGEST_COLUMN_ICON_1;
        public static final String COL_PRIORITY = "Priority";
        public static final String COL_TIME = "Time";
    }

    public static final class RuleUser implements BaseColumns {

        public static final String PATH = "rule/user";

        public static final Uri URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build();

        public static final String TYPE_DIR = ContentResolver.CURSOR_DIR_BASE_TYPE + "/" + AUTH + "/" + PATH;
        public static final String TYPE_ITEM = ContentResolver.CURSOR_ITEM_BASE_TYPE + "/" + AUTH + "/" + PATH;

        public static final String TABLE_NAME = RuleUser.class.getSimpleName();

        public static final String COL_POSITION = "Position";
        public static final String COL_EDITABLE = "Editable";
        public static final String COL_NAME = "Name";
        public static final String COL_DESCRIPTION = "Description";
        public static final String COL_MATCH = "Match";
        public static final String COL_REPLACE = "Replace";

        public static int clear(Context context) {
            ContentResolver cr = context.getContentResolver();
            return cr.delete(URI, null, null);
        }

        public static Uri insert(Context context, int position, MyCommands.RERule rule) {
            ContentResolver cr = context.getContentResolver();
            ContentValues values = new ContentValues();
            values.put(COL_POSITION, position);
            values.put(COL_EDITABLE, rule.getEditable());
            values.put(COL_NAME, rule.getName());
            values.put(COL_DESCRIPTION, rule.getDescription());
            values.put(COL_MATCH, rule.getMatch());
            values.put(COL_REPLACE, rule.getReplace());
            return cr.insert(URI, values);
        }

        public static void save(Context context, List<MyCommands.RERule> rules) {
            for(int i = 0; i < rules.size(); ++i)
                insert(context, i, rules.get(i));
        }

        public static boolean load(Context context, List<MyCommands.RERule> rules) {
            ContentResolver cr = context.getContentResolver();
            Cursor c = cr.query(URI, null, null, null, " " + COL_POSITION + " ASC ");
            if(c == null)
                return false;
            boolean empty = !c.moveToFirst();
            if(empty) {
                c.close();
                return true;
            }
            do {
                rules.add(
                        new MyCommands.RERule(
                                0 < c.getInt(c.getColumnIndex(COL_EDITABLE)),
                                c.getString(c.getColumnIndex(COL_NAME)),
                                c.getString(c.getColumnIndex(COL_DESCRIPTION)),
                                c.getString(c.getColumnIndex(COL_MATCH)),
                                c.getString(c.getColumnIndex(COL_REPLACE))
                        )
                );
            }
            while(c.moveToNext());
            c.close();
            return true;
        }


    }

}

File: myintent/src/main/java/pl/mareklangiewicz/myintent/MIContract.kt
package pl.mareklangiewicz.myintent

import android.app.SearchManager
import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import pl.mareklangiewicz.myutils.RERule
import pl.mareklangiewicz.myutils.kclear
import pl.mareklangiewicz.myutils.kinsert
import pl.mareklangiewicz.myutils.kquery
import pl.mareklangiewicz.myutils.str
import pl.mareklangiewicz.myutils.kgetString
import pl.mareklangiewicz.myutils.kgetInt

/**
 * Created by Marek Langiewicz on 07.10.15.
 */
object MIContract {

    val AUTH = BuildConfig.AUTHORITY
    val BASE_URI = Uri.parse("content://$AUTH")!!

    object CmdRecent {

        val PATH = "cmd/recent"

        val URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build()!!

        val TYPE_DIR = "${ContentResolver.CURSOR_DIR_BASE_TYPE}/$AUTH/$PATH"
        val TYPE_ITEM = "${ContentResolver.CURSOR_ITEM_BASE_TYPE}/$AUTH/$PATH"

        val TABLE_NAME = CmdRecent::class.java.simpleName!!

        val COL_COMMAND = "Command"
        val COL_TIME = "Time"

        fun clear(context: Context): Int = context.contentResolver.kclear(URI)

        fun insert(context: Context, command: String): Uri = context.contentResolver.kinsert(URI) {
            put(COL_COMMAND, command)
            put(COL_TIME, System.currentTimeMillis())
        }

        /**
         * limit <= 0 means no limit (default is 0 - so by default there is no limit clause)
         * returns true if success
         */
        fun load(context: Context, commands: MutableList<String>, limit: Int = 0): Boolean
                = context.contentResolver.kquery(
                uri = URI,
                sortOrder = " $COL_TIME DESC ${if (limit <= 0) "" else " LIMIT ${limit.str} "}"
        ) {
            if (it.moveToFirst())
                do commands.add(it.kgetString(COL_COMMAND))
                while (it.moveToNext())
            true
        }
    }


    object CmdExample {

        val PATH = "cmd/example"

        val URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build()!!

        val TYPE_DIR = "${ContentResolver.CURSOR_DIR_BASE_TYPE}/$AUTH/$PATH"
        val TYPE_ITEM = "${ContentResolver.CURSOR_ITEM_BASE_TYPE}/$AUTH/$PATH"

        val TABLE_NAME = CmdExample::class.java.simpleName!!

        val COL_COMMAND = "Command"
        val COL_PRIORITY = "Priority"

        fun clear(context: Context): Int = context.contentResolver.kclear(URI)

        fun insert(context: Context, command: String, priority: Long): Uri = context.contentResolver.kinsert(URI) {
            put(COL_COMMAND, command)
            put(COL_PRIORITY, priority)
        }

        fun load(context: Context, commands: MutableList<String>): Boolean = context.contentResolver.kquery(
                uri = URI,
                sortOrder = " $COL_PRIORITY DESC "
        ) {
            if (it.moveToFirst()) do commands.add(it.kgetString(COL_COMMAND))
            while (it.moveToNext())
            true
        }
    }


    object CmdSuggest {

        val PATH = "cmd/${SearchManager.SUGGEST_URI_PATH_QUERY}"

        val URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build()!!

        val TYPE_DIR = "${ContentResolver.CURSOR_DIR_BASE_TYPE}/$AUTH/$PATH"

        val TABLE_NAME = CmdSuggest::class.java.simpleName!!

        val COL_QUERY = SearchManager.SUGGEST_COLUMN_QUERY
        val COL_TEXT = SearchManager.SUGGEST_COLUMN_TEXT_1
        val COL_ICON = SearchManager.SUGGEST_COLUMN_ICON_1
        val COL_PRIORITY = "Priority"
        val COL_TIME = "Time"
    }


    object RuleUser {

        val PATH = "rule/user"

        val URI = BASE_URI.buildUpon().appendEncodedPath(PATH).build()!!

        val TYPE_DIR = "${ContentResolver.CURSOR_DIR_BASE_TYPE}/$AUTH/$PATH"
        val TYPE_ITEM = "${ContentResolver.CURSOR_ITEM_BASE_TYPE}/$AUTH/$PATH"

        val TABLE_NAME = RuleUser::class.java.simpleName!!

        val COL_POSITION = "Position"
        val COL_EDITABLE = "Editable"
        val COL_NAME = "Name"
        val COL_DESCRIPTION = "Description"
        val COL_MATCH = "Match"
        val COL_REPLACE = "Replace"

        fun clear(context: Context): Int = context.contentResolver.kclear(URI)

        fun insert(context: Context, position: Int, rule: RERule): Uri = context.contentResolver.kinsert(URI) {
            put(COL_POSITION, position)
            put(COL_EDITABLE, rule.editable)
            put(COL_NAME, rule.name)
            put(COL_DESCRIPTION, rule.description)
            put(COL_MATCH, rule.match)
            put(COL_REPLACE, rule.replace)
        }

        fun save(context: Context, rules: List<RERule>) {
            for (i in rules.indices)
                insert(context, i, rules[i])
        }

        fun load(context: Context, rules: MutableList<RERule>): Boolean = context.contentResolver.kquery(
                uri = URI,
                sortOrder = " $COL_POSITION ASC "
        ) {
            if (it.moveToFirst()) do rules.add(
                    RERule(
                            it.kgetString(COL_MATCH),
                            it.kgetString(COL_REPLACE),
                            it.kgetString(COL_NAME),
                            it.kgetString(COL_DESCRIPTION),
                            it.kgetInt(COL_EDITABLE) > 0
                    )
            )
            while (it.moveToNext())
            true
        }
    }
}
